### PR TITLE
[CALCITE-4176] Key descriptor can be optional in SESSION table function (Rui Wang)

### PR DESCRIPTION
**Pull request by amaliujia**
https://issues.apache.org/jira/browse/CALCITE-4176

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @danny0405 

Do you have any suggestions on CALCITE-4176? Espeically the idea to make the argument list of SESSION table function as: `data, timecol, key optional, size`?

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 1]
```diff
@@ -10435,32 +10435,30 @@ private void checkCustomColumnResolving(String table) {
         + "size => interval '1' hour))")
         .fails("Param 'data' not found in function 'SESSION'; did you mean 'DATA'\\?");
     sql("select * from table(\n"
-        + "^session(\n"
+        + "session(\n"
         + "data => table orders,\n"
         + "key => descriptor(productid),\n"
```
Comment (amaliujia) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:10440]: I found that optional named argument is not working here, even though I have do 

```
    @Override public boolean isOptional(int i) {
      return false;
      return i == 2;
    }
```
I expect Calcite can detect required `timecol` has missing. I haven't figured out the reason of this.

**Comment by danny0405** [comment]
Hi, @amaliujia , what is the progress of this PR ?

**Comment by amaliujia** [comment]
@danny0405 

Got really busy recently and I can try to finish this one before next release. If there is an urgency that requires this change, I might be able to pick this one earlier.

**Comment by amaliujia** [comment]
tracking in https://github.com/apache/calcite/pull/2164

**Pull request closed without merging**